src/main/java/org/jitsi/jicofo/xmpp/BaseBrewery.java

```diff
@@ -0,0 +1,380 @@
+            chatRoom.addMemberPresenceListener(this);
+            chatRoom.addMemberPropertyChangeListener(this);
+            chatRoom.join();
+
+            logger.info("Joined brewery room: " + breweryName);
+        }
+        catch (OperationFailedException | OperationNotSupportedException e)
+        {
+            logger.error("Failed to create room: " + breweryName, e);
+        }
+    }
+
+    /**
+     * Stops detector by removing listeners and leaving brewery room.
+     */
+    private void stop()
+    {
+        if (chatRoom != null)
+        {
+            chatRoom.removeMemberPresenceListener(this);
+            chatRoom.removeMemberPropertyChangeListener(this);
+            chatRoom.leave();
+            chatRoom = null;
+
+            logger.info("Left brewery room: " + breweryName);
+        }
+
+        // Clean up the list of service instances
+        List<BrewInstance> instancesCopy = new ArrayList<>(instances);
+        for (BrewInstance i : instancesCopy)
+        {
+            removeInstance(i);
+        }
+    }
+
+    @Override
+    synchronized public void memberPresenceChanged(
+        ChatRoomMemberPresenceChangeEvent presenceEvent)
+    {
+        XmppChatMember chatMember
+            = (XmppChatMember) presenceEvent.getChatRoomMember();
+        String eventType = presenceEvent.getEventType();
+        if (ChatRoomMemberPresenceChangeEvent.MEMBER_JOINED.equals(eventType))
+        {
+            // Process idle or busy
+            processMemberPresence(chatMember);
+        }
+        else if (ChatRoomMemberPresenceChangeEvent.MEMBER_LEFT.equals(eventType)
+            || ChatRoomMemberPresenceChangeEvent.MEMBER_KICKED.equals(eventType)
+            || ChatRoomMemberPresenceChangeEvent.MEMBER_QUIT.equals(eventType))
+        {
+            // Process offline status
+            BrewInstance instance = find(chatMember.getContactAddress());
+
+            if (instance != null)
+                removeInstance(instance);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    synchronized public void chatRoomPropertyChanged(
+        ChatRoomMemberPropertyChangeEvent memberPropertyEvent)
+    {
+        XmppChatMember member
+            = (XmppChatMember) memberPropertyEvent.getSourceChatRoomMember();
+
+        processMemberPresence(member);
+    }
+
+    /**
+     * Process chat room member status changed. Extract the appropriate
+     * presence extension and use it to process it further.
+     * @param member the chat member to process
+     */
+    private void processMemberPresence(XmppChatMember member)
+    {
+        Presence presence = member.getPresence();
+
+        if (presence == null)
+            return;
+
+        PacketExtension ext
+            = presence.getExtension(extensionElementName, extensionNamespace);
+
+        // if the extension is required and we do not have it skip processing
+        if (ext == null && requiredExtension)
```

Comment:
If it's a problem to have it added on join it will appear once it's added (presence will be updated on extension add)